src/lib/cache/SubscriptionCache.cpp

```diff
@@ -0,0 +1,1000 @@
+*
+* semInit - 
+*/
+void SubscriptionCache::semInit(void)
+{
+  sem_init(&mutex, 0, 1); // Shared between threads, not taken initially
+}
+
+
+
+/* ****************************************************************************
+*
+* semTake - 
+*/
+void SubscriptionCache::semTake(void)
+{
+  if (pthread_self() == mutexOwner)
+  {
+    return;
+  }
+
+  struct timespec  startTime;
+  struct timespec  endTime;
+  struct timespec  diffTime;
+
+  if (semTimeStatistics)
+  {
+    clock_gettime(CLOCK_REALTIME, &startTime);
+  }
+
+  sem_wait(&mutex);
+  mutexOwner = pthread_self();
+
+  if (semTimeStatistics)
+  {
+    clock_gettime(CLOCK_REALTIME, &endTime);
+    clock_difftime(&endTime, &startTime, &diffTime);
+    clock_addtime(&subCacheMutexTime, &diffTime);
+  }
+}
+
+
+/* ****************************************************************************
+*
+* semGive - 
+*/
+void SubscriptionCache::semGive(void)
+{
+  sem_post(&mutex);
+  mutexOwner = 0;
+}
+
+
+
+/* ****************************************************************************
+*
+* lookup - 
+*/
+void SubscriptionCache::lookup
+(
+  const std::string&           tenant,
+  const std::string&           servicePath,
+  const std::string&           id,
+  const std::string&           type,
+  const std::string&           attributeName,
+  std::vector<Subscription*>*  subV
+)
+{
+  LM_M(("KZ: Looking up a sub in a cache of %d subs (tenant='%s',spath='%s')", subs.size(), tenant.c_str(), servicePath.c_str()));
+
+  for (unsigned int ix = 0; ix < subs.size(); ++ix)
+  {
+    if (subs[ix]->match(tenant, servicePath, id, type, attributeName))
+    {
+      subV->push_back(subs[ix]);
+    }
+  }
+
+  ++noOfSubCacheLookups;
+  LM_M(("KZ: returning a list of %d Subscriptions for the lookup", subV->size()));
+}
+
+
+
+/* ****************************************************************************
+*
+* lookupById - 
+*/
+Subscription* SubscriptionCache::lookupById
+(
+  const std::string&  tenant,
+  const std::string&  servicePath,
+  const std::string&  subId
+)
+{
+  unsigned int ix;
+
+  for (ix = 0; ix < subs.size(); ++ix)
+  {
+    if (subs[ix]->tenant != tenant)
+    {
+      continue;
+    }
+
+    if (subs[ix]->servicePathMatch(servicePath) == false)
+    {
+      continue;
+    }
+
+    if (subs[ix]->subscriptionId != subId)
+    {
+      continue;
+    }
+
+    return subs[ix];
+  }
+
+  return NULL;
+}
+
+
+
+/* ****************************************************************************
+*
+* subTreat - 
+*/
+static void subTreat(std::string tenant, BSONObj& bobj)
+{
+  BSONElement  idField = bobj.getField("_id");
+
+  if (idField.eoo() == true)
+  {
+    LM_E(("Database Error (error retrieving _id field in doc: '%s')", bobj.toString().c_str()));
+    return;
+  }
+
+
+
+  //
+  // 01. Extract values from database object 'bobj'
+  //
+
+  std::string               subId         = idField.OID().toString();
+  int64_t                   expiration    = bobj.getField("expiration").Long();
```

Comment:
To keep synzed DB and cache, please use the CSUB_ defines (where applicable) defined in MongoGlobal.h.